### PR TITLE
Add finetuning generation

### PR DESCRIPTION
**Pull request by gkumbhat**
closes: #71 

### Changes

- Add fine-tuning module that uses HF Trainer to showcase how we can use it to fine-tune a model.

### Notes:
- This is purposefully only using 1 type of trainer and supporting 1 type of models. The purpose is to lay out small changes and iteratively improve to avoid giant PRs.
- For this PR to work fully, we need to update `transformers` version to include the bugfix merged in the PR https://github.com/huggingface/transformers/pull/24759

### CODE REVIEW

File: tests/modules/text_generation/test_fine_tuning.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            [
+                GenerationTrainRecord(
+                    input="@foo what a cute dog!", output="no complaint"
+                ),
+                GenerationTrainRecord(
+                    input="@bar this is the worst idea ever.", output="complaint"
+                ),
+            ]
+        ),
+        "torch_dtype": torch.bfloat16,
```
Comment (alex-jw-brooks) [inline tests/modules/text_generation/test_fine_tuning.py:42]: Small thing, but can we use float32 here, since bfloat16 is not supported on all devices? Unsupported type fallbacks should be handled in other tests, but it would be nice to make sure this test runs identically on most devices

File: tests/modules/text_generation/test_fine_tuning.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    model = FineTuning.train(
+        **train_kwargs
+    )
+    assert isinstance(model.model, Trainer)
+    # Ensure that we can get something out of it
+    pred = model.run("@bar what a cute cat!")
```
Comment (alex-jw-brooks) [inline tests/modules/text_generation/test_fine_tuning.py:49]: Can you add a check for the type of `model`?

File: tests/modules/text_generation/test_fine_tuning.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import torch
+from transformers import Trainer
+
+# First Party
+import caikit
+
+# Local
+from caikit_nlp.data_model import GeneratedResult, GenerationTrainRecord
+from caikit_nlp.modules.text_generation import FineTuning
+from caikit_nlp.resources.pretrained_model import HFAutoCausalLM, HFAutoSeq2SeqLM
```
Comment (alex-jw-brooks) [inline tests/modules/text_generation/test_fine_tuning.py:12]: Nit: some unused imports here, e.g., `HFAutoCausalLM`

File: caikit_nlp/modules/text_generation/fine_tuning.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+                max_new_tokens=max_new_tokens,
+                min_new_tokens=min_new_tokens,
+            )
+
+            generated_text = self.tokenizer.batch_decode(
+                generated_tokens.detach().cpu().numpy(), skip_special_tokens=True
+            )
+
+        else:
+            raise NotImplementedError(
```
Comment (alex-jw-brooks) [inline caikit_nlp/modules/text_generation/fine_tuning.py:220]: Error types should be wrapped by the error handler!

File: caikit_nlp/modules/text_generation/fine_tuning.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+from caikit.core.toolkit import error_handler, wip_decorator
+import alog
+
+# Local
+from ...data_model import GeneratedResult, GenerationTrainRecord
+from ...toolkit.data_stream_wrapper import SimpleIterableStreamWrapper
+from ...toolkit.data_type_utils import get_torch_dtype
+from .text_generation_task import TextGenerationTask
+
+log = alog.use_channel("TXT_GEN")
```
Comment (alex-jw-brooks) [inline caikit_nlp/modules/text_generation/fine_tuning.py:39]:
```suggestion
log = alog.use_channel("FIN_TUN_GEN")
```

So that we don't have collisions in logging channels with the separate text generation module

File: caikit_nlp/modules/text_generation/fine_tuning.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        # available in different devices (and configuration), depending on
+        # how it was trained. For now (July 10, 2023), we are not trying to
+        # extract the model out from trainer itself, since that would require
+        # us to essentially save it or reconstruct it to do normal inferring.
+        self.model = model
+
+    @classmethod
+    def train(
+        cls,
+        base_model: str,  # TODO: Union[str, PretrainedModelBase]
```
Comment (alex-jw-brooks) [inline caikit_nlp/modules/text_generation/fine_tuning.py:71]: Somewhat unrelated - I can't remember the exact reason unions were disallowed in our type hints for a little while, maybe something for caikit runtime type hint inspection and unions being unsupported at the time. I think that some work has been going on around that recently though - do you know if we can use them yet?

File: caikit_nlp/modules/text_generation/fine_tuning.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    ):
+        """
+        # FIXME: Below is currently configured for Seq2Seq only
+        """
+
+        torch_dtype = get_torch_dtype(torch_dtype)
+
+        ## Load base model
+        if isinstance(base_model, str):
+            model_config = AutoConfig.from_pretrained(base_model)
```
Comment (alex-jw-brooks) [inline caikit_nlp/modules/text_generation/fine_tuning.py:91]: Hmm - this is duplicated in a few places, e.g., [here](https://github.com/caikit/caikit-nlp/blob/main/caikit_nlp/modules/text_generation/text_generation.py#L119). Can you add a note to consolidate these into a common base class in the future or something? Would like to try to keep tabs on some of these things for tech health

File: caikit_nlp/modules/text_generation/fine_tuning.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        cls,
+        base_model: str,  # TODO: Union[str, PretrainedModelBase]
+        train_stream: DataStream[GenerationTrainRecord],
+        torch_dtype: str = None,  # TODO: Optional[Union[torch.dtype, str]]
+        max_source_length: int = 256,
+        max_target_length: int = 128,
+        batch_size: int = 8,
+        num_epochs: int = 5,
+        accumulate_steps: int = 32,
+        shuffle: bool = True,
```
Comment (alex-jw-brooks) [inline caikit_nlp/modules/text_generation/fine_tuning.py:79]: Should `shuffle` really be parameterized here? Might be better to just always use True for this. E.g., this is what we do in the wrapper for Peft prompt tuning ([here](https://github.com/caikit/caikit-nlp/blob/main/caikit_nlp/modules/text_generation/peft_prompt_tuning.py#L730))

File: caikit_nlp/modules/text_generation/fine_tuning.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+            max_target_length=max_target_length,
+            shuffle=shuffle,
+        )
+
+        ## TODO: Fetch trainer from resource
+
+        # TODO: Make this whole thing configurable by end-users, by optionally accepting `training_args`
+        # as argument to this train function.
+        # TODO: Remove all the default used below and make them all configurable
+        training_args = Seq2SeqTrainingArguments(
```
Comment (alex-jw-brooks) [inline caikit_nlp/modules/text_generation/fine_tuning.py:123]: I think I might misunderstand this - is this module not providing fine tuning support to causal LM also? Also, can we fallback to defaults in the trainer or something instead of hard coding default overrides here?

**Review by alex-jw-brooks — COMMENTED** [review]
Thanks Gaurav! Some initial thoughts

File: caikit_nlp/modules/text_generation/fine_tuning.py
[the remark this answers is not in the document]
Comment (gkumbhat) [reply caikit_nlp/modules/text_generation/fine_tuning.py:1]: agreed, this is only temporarily a separate module. I was imaging it to go into `.train` for the `text_generation` module and a lot of the core functionality go in toolkit or something. Like was planning to add a `get_trainer` function in individual resources which would move a lot of the code to individual resources so we would be able to basically use same logic in different `task` specific modules

File: caikit_nlp/modules/text_generation/fine_tuning.py (hunk 6 above)
Comment (gkumbhat) [reply caikit_nlp/modules/text_generation/fine_tuning.py:71]: hmm. honestly not sure, I guess certainly something we could try, but there is a lot of things influx in data models, so we might want things to be more stable and then make these changes

File: caikit_nlp/modules/text_generation/fine_tuning.py (hunk 8 above)
Comment (gkumbhat) [reply caikit_nlp/modules/text_generation/fine_tuning.py:79]: hmm. good question, I was thinking might be just better to provide option, but may be doesn't need to be 🤔

File: caikit_nlp/modules/text_generation/fine_tuning.py (hunk 7 above)
Comment (gkumbhat) [reply caikit_nlp/modules/text_generation/fine_tuning.py:91]: good idea. Will do that

File: caikit_nlp/modules/text_generation/fine_tuning.py (hunk 9 above)
Comment (gkumbhat) [reply caikit_nlp/modules/text_generation/fine_tuning.py:123]: this will eventually support trainings for various types of models, for this iteration, I just started with seq2seq. In next iteration, I'll move this whole logic to  individual `resources`, so we just do `.get_trainer()` on it.

I left a comment related to that [here](https://github.com/caikit/caikit-nlp/pull/72/files#diff-5e944359dd88fad4652c05a882bc242a6b567ded247a67ac24c9fc71ef9c263bR118)

File: caikit_nlp/toolkit/data_stream_wrapper.py [hunk 10]
```diff
@@ -48,10 +50,26 @@ def __init__(self, stream, shuffle, buffer_size=None):
...
+        # each worker
+        else:  # in a worker process
+            # split workload
+            per_worker = int(
+                math.ceil((self.end - self.start) / float(worker_info.num_workers))
+            )
+            worker_id = worker_info.id
+            iter_start = self.start + worker_id * per_worker
+            iter_end = min(iter_start + per_worker, self.end)
+        return iter(range(iter_start, iter_end))
```
[the remark this answers is not in the document]
Comment (gkumbhat) [reply caikit_nlp/toolkit/data_stream_wrapper.py:72]: you mean add for multi-device case? that part is tricky since we won't have multi-device available in testing environment. I guess, I can patch the `get_worker_info` from transformers to write the test for this 🤔

File: tests/modules/text_generation/test_fine_tuning.py (hunk 1 above)
Comment (gkumbhat) [reply tests/modules/text_generation/test_fine_tuning.py:42]: good point!

File: tests/modules/text_generation/test_fine_tuning.py (hunk 2 above)
Comment (gkumbhat) [reply tests/modules/text_generation/test_fine_tuning.py:49]: you mean check for `model.model` ?

File: caikit_nlp/modules/text_generation/fine_tuning.py (hunk 8 above)
Comment (gkumbhat) [reply caikit_nlp/modules/text_generation/fine_tuning.py:79]: changed to  always be `True`

File: caikit_nlp/modules/text_generation/fine_tuning.py
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply caikit_nlp/modules/text_generation/fine_tuning.py:1]: Awesome, I think that sounds like a good plan

**Review by alex-jw-brooks — APPROVED** [review]
Approving for now - I think the longer term plan for this sounds good, and it's fine to keep up while it's in wip state. Thanks @gkumbhat!
